File: projects/visionary/upstream/src/rate-limiter.js
// Rate limiter — per-agent token bucket.
//
// Single-process, in-memory. Config is loaded from the `settings` table on
// first use and written back on configure(). Refill happens lazily on acquire()
// rather than via a timer, so no background interval is needed.

const DEFAULT_CAPACITY = 10;
const DEFAULT_REFILL_PER_SECOND = 1;

// buckets: agentId -> { tokens, capacity, refillPerSecond, lastRefillMs }
const buckets = new Map();

// Injected on first acquire() when stmts are provided.
let _stmts = null;

function _loadConfig(agentId) {
  if (!_stmts) return null;
  try {
    const row = _stmts.getSettings.get('rate_limit:' + agentId);
    if (!row) return null;
    return JSON.parse(row.value_json);
  } catch { return null; }
}

function _saveConfig(agentId, capacity, refillPerSecond) {
  if (!_stmts) return;
  try {
    _stmts.upsertSettings.run({
      key: 'rate_limit:' + agentId,
      value_json: JSON.stringify({ capacity, refill_per_second: refillPerSecond })
    });
  } catch { /* non-fatal — config stays in memory */ }
}

function _getBucket(agentId) {
  if (!buckets.has(agentId)) {
    const cfg = _loadConfig(agentId) || {};
    const capacity = cfg.capacity || DEFAULT_CAPACITY;
    const refillPerSecond = cfg.refill_per_second || DEFAULT_REFILL_PER_SECOND;
    buckets.set(agentId, {
      tokens: capacity,
      capacity,
      refillPerSecond,
      lastRefillMs: Date.now()
    });
  }
  return buckets.get(agentId);
}

function _applyRefill(bucket) {
  const now = Date.now();
  const elapsed = (now - bucket.lastRefillMs) / 1000;
  const added = elapsed * bucket.refillPerSecond;
  if (added >= 1) {
    bucket.tokens = Math.min(bucket.capacity, bucket.tokens + Math.floor(added));
    bucket.lastRefillMs = now;
  }
}

/**
 * Inject the DB statement bundle so config can be persisted.
 * Call once at startup from server.js or wherever executeWithFailover is wired.
 */
function init(stmts) {
  _stmts = stmts;
}

/**
 * Try to consume one token for agentId.
 * Returns true if the dispatch is allowed, false if the bucket is empty.
 * Does NOT block — callers decide what to do on false.
 */
function acquire(agentId) {
  const bucket = _getBucket(agentId);
  _applyRefill(bucket);
  if (bucket.tokens < 1) return false;
  bucket.tokens -= 1;
  return true;
}

/**
 * Manually trigger a refill pass across all known buckets.
 * Useful for tests or an external keepalive interval.
 */
function refill() {
  for (const bucket of buckets.values()) {
    _applyRefill(bucket);
  }
}

/**
 * Override per-agent capacity and refill rate.
 * Resets the bucket to full capacity with the new settings.
 */
function configure(agentId, capacity, refillPerSecond) {
  capacity = Math.max(1, capacity || DEFAULT_CAPACITY);
  refillPerSecond = Math.max(0.01, refillPerSecond || DEFAULT_REFILL_PER_SECOND);
  buckets.set(agentId, {
    tokens: capacity,
    capacity,
    refillPerSecond,
    lastRefillMs: Date.now()
  });
  _saveConfig(agentId, capacity, refillPerSecond);
}

/**
 * Return a snapshot of the bucket state (for /api/agents/:id/throttle or tests).
 */
function status(agentId) {
  const bucket = _getBucket(agentId);
  _applyRefill(bucket);
  return {
    tokens: bucket.tokens,
    capacity: bucket.capacity,
    refill_per_second: bucket.refillPerSecond
  };
}

module.exports = { init, acquire, refill, configure, status };

File: projects/visionary/upstream/src/visionary/orchestration/rate_limiter.py
# src/visionary/orchestration/rate_limiter.py
"""Per-agent token bucket throttle.

Config persists in `settings` under key `rate_limit:<agent_id>`. Counters
are in-memory (single-user, single-process). On `RateLimiter()` construction
we lazily hydrate caps from settings on first use; defaults if no setting.
"""

import json
import time
from typing import Any

from visionary.db.database import Database
from visionary.db.statements import Statements

DEFAULT_CAPACITY = 10
DEFAULT_REFILL_PER_SECOND = 1.0


def _key(agent_id: str) -> str:
    return f"rate_limit:{agent_id}"


class _Bucket:
    __slots__ = ("capacity", "refill_per_second", "tokens", "last_refill_ts")

    def __init__(self, capacity: int, refill_per_second: float):
        self.capacity = capacity
        self.refill_per_second = refill_per_second
        self.tokens = float(capacity)
        self.last_refill_ts = time.monotonic()


class RateLimiter:
    def __init__(self, db: Database):
        self._db = db
        self._stmts = Statements(db)
        self._buckets: dict[str, _Bucket] = {}

    def _get_or_load(self, agent_id: str) -> _Bucket:
        if agent_id not in self._buckets:
            row = self._stmts.get_setting(_key(agent_id))
            if row is not None:
                cfg = json.loads(row["value_json"])
                cap = int(cfg.get("capacity", DEFAULT_CAPACITY))
                rate = float(cfg.get("refill_per_second", DEFAULT_REFILL_PER_SECOND))
            else:
                cap = DEFAULT_CAPACITY
                rate = DEFAULT_REFILL_PER_SECOND
            self._buckets[agent_id] = _Bucket(cap, rate)
        return self._buckets[agent_id]

    def _refill(self, bucket: _Bucket) -> None:
        now = time.monotonic()
        elapsed = now - bucket.last_refill_ts
        if elapsed <= 0:
            return
        bucket.tokens = min(bucket.capacity, bucket.tokens + elapsed * bucket.refill_per_second)
        bucket.last_refill_ts = now

    def acquire(self, agent_id: str) -> bool:
        bucket = self._get_or_load(agent_id)
        self._refill(bucket)
        if bucket.tokens >= 1.0:
            bucket.tokens -= 1.0
            return True
        return False

    def configure(self, agent_id: str, capacity: int, refill_per_second: float) -> None:
        self._stmts.upsert_setting(
            _key(agent_id),
            json.dumps({"capacity": capacity, "refill_per_second": refill_per_second}),
        )
        bucket = _Bucket(capacity, refill_per_second)
        self._buckets[agent_id] = bucket

    def status(self, agent_id: str) -> dict[str, Any]:
        bucket = self._get_or_load(agent_id)
        self._refill(bucket)
        return {
            "capacity": bucket.capacity,
            "refill_per_second": bucket.refill_per_second,
            "tokens": bucket.tokens,
        }
